python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,322 @@
+            Defaults to False. See Appendix A.1 here
+            https://arxiv.org/pdf/1711.09846.pdf.
+
+    Example:
+        >>> pb2 = PB2(
+        >>>     time_attr="timesteps_total",
+        >>>     metric="episode_reward_mean",
+        >>>     mode="max",
+        >>>     perturbation_interval=10000,
+        >>>     hyperparam_mutations={
+        >>>         # These must be continuous, currently a limitation.
+        >>>         "factor_1": lambda: random.uniform(0.0, 20.0),
+        >>>     })
+        >>> tune.run({...}, num_samples=8, scheduler=pb2)
+    """
+
+    def __init__(self,
+                 time_attr: str = "time_total_s",
+                 reward_attr: Optional[str] = None,
+                 metric: Optional[str] = None,
+                 mode: Optional[str] = None,
+                 perturbation_interval: float = 60.0,
+                 hyperparam_mutations: Dict = None,
+                 quantile_fraction: float = 0.25,
+                 log_config: bool = True,
+                 require_attrs: bool = True,
+                 synch: bool = False):
+
+        resample_probability = 0
+        custom_explore_fn = None
+
+        super(PopulationBasedTraining, self).__init__(
+            self, time_attr, reward_attr, metric, mode, perturbation_interval,
+            hyperparam_mutations, quantile_fraction, resample_probability,
+            custom_explore_fn, log_config, require_attrs, synch)
+
+        self._name = "pb2"
+
+        self.latest = 0  # when we last explored
+        self.data = pd.DataFrame()
+
+        self.bounds = {}
+        for key, distribution in self._hyperparam_mutations.items():
+            self.bounds[key] = [
+                np.min([distribution() for _ in range(999999)]),
+                np.max([distribution() for _ in range(999999)])
+            ]
+
+    def _save_trial_state(self, state, time, result, trial):
+        score = super(PopulationBasedTraining, self)._save_trial_state(
+            state, time, result, trial)
+
+        # Data logging for PB2.
+
+        # Collect hyperparams names and current values for this trial.
+        names = []
+        values = []
+        for key in self._hyperparam_mutations.keys():
+            names.append(str(key))
+            values.append(trial.config[key])
+
+        # Store trial state and hyperparams in dataframe.
+        # this needs to be made more general.
+        lst = [[trial, result[self._time_attr]] + values + [score]]
+        cols = ['Trial', 'T'] + names + ['Reward']
+        entry = pd.DataFrame(lst, columns=cols)
+
+        self.data = pd.concat([self.data, entry]).reset_index(drop=True)
+        self.data.Trial = self.data.Trial.astype('str')
+
+    def _get_new_config(self, trial, trial_to_clone):
+        # If we are at a new timestep, we dont want to penalise for trials
+        # still going
+        if self.data['T'].max() > self.latest:
+            self.current = None
```

Comment:
Can you set self.current = None in __init__ and also add a comment explaining what exactly self.current is/how it's different from trial?

Reply:
Done.